doc: add more literature to api doc (#308)
metal3-io/baremetal-operator: docs/api.md

two well differentiated sections, the bare metal host specification
and its current status.

### BareMetalHost spec

The *BareMetalHost's* *spec* defines the desire state of the host. It contains
mainly, but not only, provisioning details.

#### Spec fields

* *bmc* -- The connection information for the BMC (Baseboard Management
  Controller) on the host.
  * *address* -- The URL for communicating with the BMC controller, based
    on the provider being used, the URL will look as follows:
    * IPMI
      * `ipmi://<host>:<port>`, an unadorned `<host>:<port>` is also accepted
        and the port is optional, if using the default one (623).
    * Dell iDRAC
      * `idrac://` (or `idrac+http://` to disable TLS).
    * Fujitsu iRMC
      * `irmc://<host>:<port>`, where `<port>` is optional if using the default.
    * Redfish
      * `redfish://` (or `redfish+http://` to disable TLS), the hostname
        or IP address, and the path to the system ID are required,
        for example `redfish://myhost.example/redfish/v1/Systems/MySystemExample`

  * *credentialsName* -- A reference to a *secret* containing the connection

>     details, at least username and password, for the BMC.

Reviewer: ```suggestion
    username and password for the BMC.
```

We only want the username and password, so saying "at least" is confusing here.

* *online* -- A boolean indicating whether the host should be powered on
  (true) or off (false). Changing this value will trigger a change in
  power state on the physical host.

* *consumerRef* -- A reference to another resource that is using the
  host, it could be empty if the host is not being currently used.
  For example, a *Machine* resource when the host is being used by the

>   [*machine-api*](https://github.com/openshift/machine-api-operator).

Reviewer: ```suggestion
  [*machine-api*](https://github.com/kubernetes-sigs/cluster-api).
```

We should link to the upstream definition of the machine-api.

* *externallyProvisioned* -- A boolean indicating whether the host
  provisioning and deprovisioning are managed externally. When set,

>   if no image is provided, the host's power status and hardware

Reviewer: ```suggestion
  the host's power status and hardware
```

It looks like I missed a doc update when changing the way this flag works. If it is set, the image isn't checked at all.

  inventory will be monitored but no provisioning or deprovisioning
  operations are performed on the host.

* *image* -- Holds details for the image to be deployed on a given host.
  * *url* -- The URL of an image to deploy to the host.
  * *checksum* -- The actual checksum or a URL to a file containing
    the checksum for the image at *image.url*.

> * *userData* -- A reference to the Secret containing the user provided data

Reviewer: ```suggestion
* *userData* -- A reference to the Secret containing the cloudinit user data
```

It's not "user provided data" it's the user data value for cloudinit.

Author: @dhellmann I used a kinda generic term cause I assumed it could be other technologies, like for instance ignition, is not that the case?

Reviewer: For RHCOS the data is consumed by Ignition, but it's passed in cloudinit format.

  and its namespace, so it can be attached to the host before it boots
  for configuring different aspects of the OS (like networking, storage, ...).

* *description* -- A human-provided string to help identify the host.

### BareMetalHost status

Moving onto the next block, the *BareMetalHost's* *status* which represents
the host's current state. Including tested credentials, current hardware
details, etc.

#### Status fields

* *goodCredentials* -- A reference to the secret and its namespace

>   holding the last set of credentials the system was able to validate

Reviewer: ```suggestion
  holding the last set of BMC credentials the system was able to validate
```

  as working.

* *triedCredentials* -- A reference to the secret and its namespace

>   holding the last set of credentials that were sent to

Reviewer: ```suggestion
  holding the last set of BMC credentials that were sent to
```

  the provisioning backend.

* *lastUpdated* -- The timestamp of the last time the status of the
  host was updated.

* *operationalStatus* -- The status of the server. Value is one of the
  following:
  * *OK* -- Indicates all the details for the host are known and working,
    meaning the host is correctly configured and manageable.
  * *discovered* -- Implies some of the host's details are either

>     not working correctly missing. For exmample, the BMC address is known

Reviewer: ```suggestion
    not working correctly or missing. For example, the BMC address is known
```

    * *inspecting* -- The hardware details for the host are being collected
      by an agent.
    * *power management error* -- An error was found while trying to power
      the host either on or off.

### BareMetalHost Example

The following is a complete example from a running cluster of a *BareMetalHost*
resource (in YAML), it includes its specification and status sections:

```yaml
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  creationTimestamp: "2019-09-20T06:33:35Z"
  finalizers:
  - baremetalhost.metal3.io
  generation: 2
  name: bmo-master-0
  namespace: bmo-project
  resourceVersion: "22642"
  selfLink: /apis/metal3.io/v1alpha1/namespaces/bmo-project/baremetalhosts/bmo-master-0
  uid: 92b2f77a-db70-11e9-9db1-525400764849
spec:
  bmc:
    address: ipmi://10.10.57.19
    credentialsName: bmo-master-0-bmc-secret
  bootMACAddress: 98:03:9b:61:80:48
  consumerRef:
    apiVersion: machine.openshift.io/v1beta1
    kind: Machine
    name: bmo-master-0
    namespace: bmo-project
  externallyProvisioned: true
  hardwareProfile: default
  image:
    checksum: http://172.16.1.100/images/myOS.qcow2.md5sum

>     url: http://172.16.1.100/images/myOSv1/myOS.qcow2

Reviewer: The paths in these values don't match. Should they?

Author: I guess they don't have to, but it was my intention that they'd match, fixing it, thanks!
